cache/remotecache/s3/s3.go

```diff
@@ -0,0 +1,541 @@
+	}
+	desc := ocispecs.Descriptor{
+		MediaType:   l.Annotations.MediaType,
+		Digest:      l.Blob,
+		Size:        l.Annotations.Size,
+		Annotations: annotations,
+	}
+	return &v1.DescriptorProviderPair{
+		Descriptor: desc,
+		Provider:   &s3Provider{i.cache, i.config},
+	}, nil
+}
+
+func (i *importer) load() (*v1.CacheChains, error) {
+	var config v1.CacheConfig
+	found, err := i.cache.getManifest(manifestKey(i.config, i.config.Names[0]), &config)
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return v1.NewCacheChains(), nil
+	}
+
+	allLayers := v1.DescriptorProvider{}
+
+	for _, l := range config.Layers {
+		dpp, err := i.makeDescriptorProviderPair(l)
+		if err != nil {
+			return nil, err
+		}
+		allLayers[l.Blob] = *dpp
+	}
+
+	cc := v1.NewCacheChains()
+	if err := v1.ParseConfig(config, allLayers, cc); err != nil {
+		return nil, err
+	}
+	return cc, nil
+}
+
+func (i *importer) Resolve(ctx context.Context, _ ocispecs.Descriptor, id string, w worker.Worker) (solver.CacheManager, error) {
+	cc, err := i.load()
+	if err != nil {
+		return nil, err
+	}
+
+	keysStorage, resultStorage, err := v1.NewCacheKeyStorage(cc, w)
+	if err != nil {
+		return nil, err
+	}
+
+	return solver.NewCacheManager(ctx, id, keysStorage, resultStorage), nil
+}
+
+type s3Provider struct {
+	cache  *cache
+	config *Config
+}
+
+func (p *s3Provider) ReaderAt(ctx context.Context, desc ocispecs.Descriptor) (content.ReaderAt, error) {
+	readerAtCloser := toReaderAtCloser(func(offset int64) (io.ReadCloser, error) {
+		return p.cache.getReader(blobKey(p.config, desc.Digest))
+	})
+	return &readerAt{ReaderAtCloser: readerAtCloser, desc: desc}, nil
+}
+
+type readerAt struct {
+	ReaderAtCloser
+	desc ocispecs.Descriptor
+}
+
+func (r *readerAt) Size() int64 {
+	return r.desc.Size
+}
+
+func oneOffProgress(ctx context.Context, id string) func(err error) error {
+	pw, _, _ := progress.NewFromContext(ctx)
+	now := time.Now()
+	st := progress.Status{
+		Started: &now,
+	}
+	pw.Write(id, st)
+	return func(err error) error {
+		now := time.Now()
+		st.Completed = &now
+		pw.Write(id, st)
+		pw.Close()
+		return err
+	}
+}
+
+func manifestKey(config *Config, name string) string {
+	return config.Prefix + config.ManifestsPrefix + name
+}
+
+func blobKey(config *Config, dgst digest.Digest) string {
+	return config.Prefix + config.BlobsPrefix + dgst.String()
+}
+
+type cache struct {
+	config *Config
+
+	client   *s3.S3
+	uploader *s3manager.Uploader
+}
+
+func newCache(config *Config) (*cache, error) {
+	client, err := newClient(config.Region, config.Role, config.EndpointURL, config.AccessKeyID, config.SecretAccessKey, config.S3ForcePathStyle)
+	if err != nil {
+		return nil, err
+	}
+	return &cache{
+		config: config,
+
+		client:   client,
+		uploader: s3manager.NewUploaderWithClient(client),
+	}, nil
+}
+
+func (c *cache) getManifest(key string, config *v1.CacheConfig) (bool, error) {
+	fmt.Printf("Start downloading s3://%s/%s\n", c.config.Bucket, key)
```

Comment:
Writing to stdout is not allowed in here. If you want debug logs then you need to write them to the context logger.

Reply:
removed all debug